cpp/src/parquet/geometry_statistics.h

```diff
@@ -0,0 +1,164 @@
+  EncodedGeospatialStatistics& operator=(const EncodedGeospatialStatistics&) = default;
+
+  double xmin{kInf};
+  double xmax{-kInf};
+  double ymin{kInf};
+  double ymax{-kInf};
+  double zmin{kInf};
+  double zmax{-kInf};
+  double mmin{kInf};
+  double mmax{-kInf};
+  std::vector<int32_t> geospatial_types;
+
+  bool has_z() const { return (zmax - zmin) != -kInf; }
+
+  bool has_m() const { return (mmax - mmin) != -kInf; }
+
+  bool is_set() const { return !geospatial_types.empty(); }
+};
+
+class GeospatialStatisticsImpl;
+
+/// \brief Base type for computing geospatial column statistics while writing a file
+class PARQUET_EXPORT GeospatialStatistics {
+ public:
+  GeospatialStatistics();
+  explicit GeospatialStatistics(std::unique_ptr<GeospatialStatisticsImpl> impl);
+  explicit GeospatialStatistics(const EncodedGeospatialStatistics& encoded);
+  GeospatialStatistics(GeospatialStatistics&&);
+
+  ~GeospatialStatistics();
+
+  /// \brief Return true if bounds, geometry types, and validity are identical
+  bool Equals(const GeospatialStatistics& other) const;
+
+  /// \brief Update these statistics based on previously calculated or decoded statistics
+  void Merge(const GeospatialStatistics& other);
+
+  /// \brief Update these statistics based on values
+  void Update(const ByteArray* values, int64_t num_values, int64_t null_count);
+
+  /// \brief Update these statistics based on the non-null elements of values
+  void UpdateSpaced(const ByteArray* values, const uint8_t* valid_bits,
+                    int64_t valid_bits_offset, int64_t num_spaced_values,
+                    int64_t num_values, int64_t null_count);
+
+  /// \brief Update these statistics based on the non-null elements of values
+  ///
+  /// Currently, BinaryArray and LargeBinaryArray input is supported.
+  void Update(const ::arrow::Array& values);
+
+  /// \brief Return these statistics to an empty state
+  void Reset();
+
+  /// \brief Encode the statistics for serializing to Thrift
+  ///
+  /// If invalid WKB was encountered, empty encoded statistics are returned
+  /// (such that is_set() returns false and they should not be written).
+  EncodedGeospatialStatistics Encode() const;
+
+  /// \brief Returns true if all WKB encountered was valid or false otherwise
+  bool is_valid() const;
+
+  std::shared_ptr<GeospatialStatistics> clone() const;
+
+  /// \brief Update these statistics with previously generated statistics
+  void Decode(const EncodedGeospatialStatistics& encoded);
+
+  /// \brief The minimum encountered value in the X dimension, or Inf if no X values were
+  /// encountered.
+  ///
+  /// The Parquet definition allows for "wrap around" bounds where xmin > xmax. In this
+  /// case, these bounds represent the union of the intervals [xmax, Inf] and [-Inf,
+  /// xmin]. This implementation does not yet generate these types of bounds but they may
+  /// be encountered in files written by other readers.
+  double GetXMin() const;
```

Comment:
Is it better to use `std::optional<double>` instead of returning Inf?

Reply:
It depends on whether `Inf/-Inf` is a common practice for bbox or a helper function is needed for users to understand that the bbox is empty.

BTW, I think C++ and Java implementations should be consistent: https://github.com/apache/parquet-java/pull/2971/files#diff-c47987281a33dfa16169ca9dc1e2011db9bf18bb8fbd27601207b5697d6f0e77